fix(catalog-client): stop double encoding query params

sennyeya opened the pull request:
## Hey, I just made a Pull Request!

Fixes #21717 and #21716. Basically, we were double encoding the query parameters which broke filters that contained URL encodable values, for #21717 this was an entity ref and for #21716 this was an annotation.

@awanlin 😁 

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

awanlin commented:
Just did a poke around in the Uffizzi build for this PR and it's all working great! Many thanks again @sennyeya 🚀

freben commented:
Hm. Might i trouble you to double check that there are tests where the key and/or value contain decidedly URL unsafe variables? Especially `&?`.

sennyeya commented on packages/catalog-client/src/CatalogClient.test.ts:
```diff
@@ -142,12 +142,47 @@ describe('CatalogClient', () => {
 
       server.use(
         rest.get(`${mockBaseUrl}/entities`, (req, res, ctx) => {
-          expect(decodeURIComponent(req.url.search)).toBe('?fields=a.b,%C3%B6');
```
When I originally adjusted this to use `decodeURIComponent`, it hid the underlying issue of the double encoding. The `%C3%B6` in this line. Adjusted this to be a two line test to ensure the encoding is happening and that the decoded value matches what we expect.

freben replied:
I think it would be worthwhile to parse the query using the standard URLSearchParams, and then assert that THAT results in the expected output. Comparing the query string itself doesn't test it fully.

sennyeya replied:
https://github.com/backstage/backstage/pull/21719/commits/634f62065f290b663e310615657ac847f348b7e2

Github Copilot is eerily good at the replacement logic.

sennyeya commented:
@freben Added! Went back through the other tests and something was bothering me, fixed it by removing _all_ `encodeURIComponent` in `CatalogClient` and delegating that to the library we're using in the generated client ([`uri-template`](https://www.npmjs.com/package/uri-template)) which already has a set of test cases. This should help us going forward as well as we won't have to track URI encodings (#19357). I added test cases on our side of this behavior that are a little weird -- verifying both the encoding and decoding -- that I'm not 100% sold on. I can remove/adjust those as needed.

freben reviewed (approved):
Nice!
